File: src/main/java/com/meepoffaith/hextrapats/casting/handlers/ScientificExponent.java
package com.meepoffaith.hextrapats.casting.handlers;

import at.petrak.hexcasting.api.casting.arithmetic.operator.Operator;
import at.petrak.hexcasting.api.casting.castables.Action;
import at.petrak.hexcasting.api.casting.castables.SpecialHandler;
import at.petrak.hexcasting.api.casting.eval.CastingEnvironment;
import at.petrak.hexcasting.api.casting.iota.DoubleIota;
import at.petrak.hexcasting.api.casting.iota.Iota;
import at.petrak.hexcasting.api.casting.iota.Vec3Iota;
import at.petrak.hexcasting.api.casting.math.HexPattern;
import at.petrak.hexcasting.api.casting.mishaps.MishapInvalidIota;
import com.meepoffaith.hextrapats.casting.bases.ConstMediaActionBase;
import com.meepoffaith.hextrapats.casting.bases.HexIotaStack;
import com.meepoffaith.hextrapats.init.SpecialHandlers;
import com.meepoffaith.hextrapats.util.HextraUtils;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import static at.petrak.hexcasting.common.lib.hex.HexIotaTypes.DOUBLE;
import static at.petrak.hexcasting.common.lib.hex.HexIotaTypes.VEC3;

public class ScientificExponent implements SpecialHandler{
    int exponent;

    public ScientificExponent(int exponent){
        this.exponent = exponent;
    }

    @Override
    public Action act(){
        return new InnerAction(exponent);
    }

    @Override
    public Text getName(){
        return HextraUtils.specialHandlerLang(SpecialHandlers.SCIENTIFIC_EXPONENT, exponent);
    }

    public static class InnerAction extends ConstMediaActionBase{
        public int argc = 1;
        public long mediaCost = 0L;
        int exponent;

        public InnerAction(int exponent){
            this.exponent = exponent;
        }

        @Override
        public List<? extends Iota> execute(HexIotaStack stack, CastingEnvironment ctx){
            Iota val = stack.get(0);

            if(val instanceof DoubleIota){
                return asActionResult(new DoubleIota(Operator.downcast(val, DOUBLE).getDouble() * Math.pow(10, exponent)));
            }else if(val instanceof Vec3Iota){
                return asActionResult(new Vec3Iota(Operator.downcast(val, VEC3).getVec3().multiply(Math.pow(10, exponent))));
            }else{
                throw new MishapInvalidIota(val, 0, Text.of("a number or vector"));
            }
        }
    }

    public static class ScientificExponentFactory implements Factory<ScientificExponent>{
        @Override
        public @Nullable ScientificExponent tryMatch(HexPattern pattern, CastingEnvironment env){
            String sig = pattern.anglesSignature();
            if(sig.startsWith("waqe") || sig.startsWith("wdeq")){
                boolean divide = sig.startsWith("wdeq");
                char[] chars = sig.substring(4).toCharArray();
                int exponent = 1;
                for(int i = 0; i < chars.length; i++){ //Code based on Sekhmet from Overevaluate
                    if(chars[i] != "qe".charAt((i + (divide ? 1 : 0)) % 2)){
                        return null;
                    }
                    exponent++;
                }
                return new ScientificExponent(exponent * (divide ? -1 : 1));
            }
            return null;
        }
    }
}

File: src/common/main/kotlin/com/meepoffaith/hextrapats/casting/handlers/ScientificExponent.kt
package com.meepoffaith.hextrapats.casting.handlers

import at.petrak.hexcasting.api.casting.asActionResult
import at.petrak.hexcasting.api.casting.castables.Action
import at.petrak.hexcasting.api.casting.castables.ConstMediaAction
import at.petrak.hexcasting.api.casting.castables.SpecialHandler
import at.petrak.hexcasting.api.casting.eval.CastingEnvironment
import at.petrak.hexcasting.api.casting.getNumOrVec
import at.petrak.hexcasting.api.casting.iota.Iota
import at.petrak.hexcasting.api.casting.math.HexPattern
import at.petrak.hexcasting.api.utils.asTranslatedComponent
import at.petrak.hexcasting.api.utils.lightPurple
import com.meepoffaith.hextrapats.init.SpecialHandlers
import com.meepoffaith.hextrapats.util.HextraUtils
import net.minecraft.text.Text
import net.minecraft.util.math.Vec3d
import kotlin.math.pow

class ScientificExponent(val exp: Int) : SpecialHandler{
    override fun act(): Action? {
        return InnerAction(exp)
    }

    override fun getName(): Text? {
        val num = Action.DOUBLE_FORMATTER.format(exp)
        return HextraUtils.specialHandlerLang(SpecialHandlers.SCI_EXP).asTranslatedComponent(num).lightPurple
    }

    class InnerAction(val exp: Int) : ConstMediaAction{
        override val argc = 1
        override fun execute(args: List<Iota>, env: CastingEnvironment): List<Iota> {
            val arg = args.getNumOrVec(0, argc)
            val pow = 10.0.pow(exp)
            return arg.map(
                { num -> (num * pow).asActionResult },
                { vec -> Vec3d(vec.x * pow, vec.y * pow, vec.z * pow).asActionResult }
            )
        }
    }

    class Factory : SpecialHandler.Factory<ScientificExponent> {
        override fun tryMatch(pattern: HexPattern, env: CastingEnvironment): ScientificExponent? {
            val sig = pattern.anglesSignature()
            if(sig.startsWith("waqe") || sig.startsWith("wdeq")){
                val divide = sig.startsWith("wdeq")
                val chars = sig.substring(4).toCharArray()
                var exponent = 1
                for(i in chars.indices){ //Code based on Sekhmet from Overevaluate
                    if(chars[i] != "qe"[(i + (if (divide) 1 else 0)) % 2]){
                        return null
                    }
                    exponent++
                }
                return ScientificExponent(exponent * (if (divide) -1 else 1))
            }
            return null
        }
    }
}
